contrib/internal/httputil/gen/gen.go

```diff
@@ -0,0 +1,36 @@
+package main
+
+import (
+	"html/template"
+	"os"
+)
+
+func main() {
+	interfaces := []string{"Flusher", "Pusher", "CloseNotifier", "Hijacker"}
+	var combos [][][]string
+	for pick := len(interfaces); pick > 0; pick-- {
+		combos = append(combos, combinations(interfaces, pick))
+	}
+
+	template.Must(template.ParseFiles("./gen.gohtml")).Execute(os.Stdout, map[string]interface{}{
+		"Interfaces":   interfaces,
+		"Combinations": combos,
+	})
+}
+
+func combinations(strs []string, pick int) (all [][]string) {
```

Comment:
I've spent quite a lot of time here trying to make sense of this algorithm. Even though it looks short and witty, and possibly a well known algorithm for computer science aficiandos, I'd expect a large percentage of people visiting this code to be like me and not understand what's going on without spending time.

Would it be possible to augment it with documentation describing in detail how it works? Or, if it's a common algorithm that has a Wiki page, simply to link to it?

I was also wondering if we could rename `strs` to `list`. "strs" makes me think "strings" which is unclear when reading. But maybe it's just me.